File: examples/SingletonPattern.py
#!/usr/bin/env python3

# pylint: disable = bad-whitespace
# pylint: disable = eval-used
# pylint: disable = invalid-name
# pylint: disable = missing-docstring
# pylint: disable = too-few-public-methods

# -------------------
# SingletonPattern.py
# -------------------

class A :
    pass

def test1 () :
    x = A()
    y = A()
    assert x != y





def singleton_function (c) :
    x = c()
    return lambda : x

@singleton_function
class B :
    pass

def test2 () :
    x = B()
    y = B()
    assert x == y




class singleton_class :
    def __init__ (self, c) :
        self.x = c()

    def __call__ (self) :
        return self.x

@singleton_class
class C :
    pass

def test3 () :
    x = C()
    y = C()
    assert x == y





def main () :
    for n in range(3) :
        eval("test" + str(n + 1) + "()")

if __name__ == "__main__" : # pragma: no cover
    print("SingletonPattern.py")
    main()
    print("Done.")

File: examples/SingletonPattern.java
// ---------------------
// SingletonPattern.java
// ---------------------

class Singleton {
    private static final Singleton _only = new Singleton();

    private Singleton ()
        {}

    public static Singleton only () {
        return _only;}

    public String f () {
        return "Singleton.f()";}}

class Singleton2 {
    private static Singleton2 _only;

    private Singleton2 ()
        {}

    public static Singleton2 only () {
        if (Singleton2._only == null)
            Singleton2._only = new Singleton2();
        return Singleton2._only;}

    public String f () {
        return "Singleton2.f()";}}

class Singleton3 {
    private Singleton3 ()
        {}

    private static class Holder {
        private static final Singleton3 _only = new Singleton3();}

    public static Singleton3 only () {
        return Holder._only;}

    public String f () {
        return "Singleton3.f()";}}

public final class SingletonPattern {
    public static void test () {
//      Singleton x = new Singleton();                    // error
    	assert(Singleton.only()     == Singleton.only());
    	assert(Singleton.only().f() == "Singleton.f()");}

    public static void main (String[] args) {
        System.out.println("SingletonPattern.java");
        test();
        System.out.println("Done.");}}
